pkg/image/controller/signature/signature_import_controller.go

```diff
@@ -0,0 +1,192 @@
+		AddFunc: func(obj interface{}) {
+			image := obj.(*imageapi.Image)
+			glog.V(4).Infof("Adding image %s", image.Name)
+			controller.enqueueImage(obj)
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			image := cur.(*imageapi.Image)
+			glog.V(4).Infof("Updating image %s", image.Name)
+			controller.enqueueImage(cur)
+		},
+	}, resyncInterval)
+
+	return controller
+}
+
+func (s *SignatureImportController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer s.queue.ShutDown()
+
+	if !cache.WaitForCacheSync(stopCh, s.imageHasSynced) {
+		return
+	}
+
+	glog.V(5).Infof("Starting workers")
+	for i := 0; i < workers; i++ {
+		go wait.Until(s.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.V(1).Infof("Shutting down")
+
+}
+
+func (s *SignatureImportController) worker() {
+	for {
+		if !s.work() {
+			return
+		}
+	}
+}
+
+// work returns true if the worker thread should continue
+func (s *SignatureImportController) work() bool {
+	key, quit := s.queue.Get()
+	if quit {
+		return false
+	}
+	defer s.queue.Done(key)
+
+	err := s.syncImageSignatures(key.(string))
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("error syncing image %s, it will be retried: %v", key.(string), err))
+		s.queue.AddRateLimited(key)
+		return true
+	}
+
+	s.queue.Forget(key)
+	return true
+}
+
+func (s *SignatureImportController) enqueueImage(obj interface{}) {
+	_, ok := obj.(*imageapi.Image)
+	if !ok {
+		return
+	}
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	s.queue.Add(key)
+}
+
+func (s *SignatureImportController) syncImageSignatures(key string) error {
+	glog.V(4).Infof("Initiating download of signatures for %s", key)
+	image, err := s.imageLister.Get(key)
+	if err != nil {
+		glog.V(4).Infof("Unable to get image %v: %v", key, err)
+		return err
+	}
+
+	currentSignatures, err := s.fetcher.DownloadImageSignatures(image)
+	if err != nil {
+		glog.V(4).Infof("Failed to fetch image %s signatures: %v", image.Name, err)
+		return err
+	}
+
+	// Having no signatures means no-op (we don't remove stored signatures when
+	// the sig-store no longer have them).
+	if len(currentSignatures) == 0 {
+		glog.V(4).Infof("No signatures dowloaded for %s", image.Name)
+		return nil
+	}
+
+	t, err := kapi.Scheme.DeepCopy(image)
+	if err != nil {
+		return err
+	}
+	newImage := t.(*imageapi.Image)
+
+	shouldUpdate := false
+
+	// Only add new signatures, do not override existing stored signatures as that
+	// can void their verification status.
+	for _, c := range currentSignatures {
+		found := false
+		for _, s := range newImage.Signatures {
+			if s.Name == c.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			newImage.Signatures = append(newImage.Signatures, c)
+			shouldUpdate = true
+		}
+	}
+
+	if len(newImage.Signatures) > s.signatureImportLimit {
```

Comment:
So this means you keep the old signatures.  How will admin know this is happening and react?  How do they react?

Reply:
they can prune signatures periodically with verify image signature remove. we can improve later with auto prune.